### PR TITLE
WIP: Rework of Sklearn Metrics

### LINKED ISSUE #1305: [Discussion][Metrics Package] Sklearn metrics
## Do we want to include metrics of `sklearn`?

The following constraints would apply if we had an sklearn metric interface:
* All imports would be within the class space, so sklearn would only remain an optional dependency (due to it's size)
* All metrics will be provided by sklearn interface at first, but will be removed later on (once we got a native pytorch implementation)

### The Advantages:
* We have a pretty impressive set of metrics directly in the beginning
* the maintenance effort is minimal
* they're well tested

### The disadvantages
* For every call to a metric, we need to convert tensors to numpy
* We will have to synchronise GPU(s) whenever we call a metric with inputs lying on GPU, which may be a bottleneck the users aren't aware of

### My proposal
* Add this layer in the beginning and try to replace as many of the metrics used there by our own :)
* Explicitly state the downfalls in docs (and maybe also raise a warning on that) to ensure users are aware of this problem.

Thoughts?

cc @Borda @williamFalcon @ethanwharris @PyTorchLightning/core-contributors

### PR DESCRIPTION
**Pull request by justusschock**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #1305  and is a rework of #1320 based on #1326 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/metrics/sklearn.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        Bridge between PyTorch Lightning and scikit-learn metrics
+
+        .. warning::
+            Every metric call will cause a GPU synchronization, which may slow down your code
+
+        .. note::
+            The order of targets and predictions may be different from the order typically used in PyTorch
+
+        Args:
+            metric_name: the metric name to import anc compute from scikit-learn.metrics
```
Comment (jeremyjordan) [inline pytorch_lightning/metrics/sklearn.py:25]: > anc
is this a typo? :)

Comment (justusschock) [reply pytorch_lightning/metrics/sklearn.py:25]: Yep it is :D Will fix this

File: environment.yml [hunk 2]
```diff
@@ -25,6 +25,8 @@ dependencies:
     - autopep8
     - check-manifest
     - twine==1.13.0
+    - pillow<7.0.0
+    - scikit-learn>=0.16.1
```
Comment (Borda) [inline environment.yml:29]: it shall be added also to requirements-extra

**Comment by Borda** [comment]
@williamFalcon why did you close the metrics PR? @justusschock?

**Comment by justusschock** [comment]
@Borda any idea, why scipy building fails?

**Comment by Borda** [comment]
@justusschock there is also problem with installing the minimal requirements, mind check?
@awaelchli will check the docs, thx

**Comment by awaelchli** [comment]
I'm done fixing the docs errors but can't push, permission denied. Is it a setting on this branch?
Otherwise please merge this branch
https://github.com/awaelchli/pytorch-lightning/tree/new_sklearn_metric

**Comment by justusschock** [comment]
Regarding the setup with minimal: This is strange, but It may be related to https://github.com/scipy/scipy/issues/12203

At first I thought it was because older torch had older numpy dependency, which might somehow be incompatible with current scipy build, but I checked and torch does not restrict it's numpy dependency...

**Comment by justusschock** [comment]
@peterbell10 sorry for pinging you here, but I didn't want to spam in the other issue. Do you maybe have any idea, why we cannot install prebuilt wheels here?

**Comment by peterbell10** [comment]
I'm not sure but you could try running pip with verbose output and see if it tells you.

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Add Metric Base Classes ([#1326](https://github.com/PyTorchLightning/pytorch-lightning/pull/1326), [#1877](https://github.com/PyTorchLightning/pytorch-lightning/pull/1877))
 
+- Add Metric Base Classes ([#1326](https://github.com/PyTorchLightning/pytorch-lightning/pull/1326))
+
```
Comment (Borda) [inline CHANGELOG.md:14]:
```suggestion
- Added Sklearn Metrics classes ([#1327](https://github.com/PyTorchLightning/pytorch-lightning/pull/1327))

```

**Comment by Borda** [comment]
@justusschock is it still WIP or ready for review/merge? it looks good to me :rabbit:

**Comment by justusschock** [comment]
@Borda this should be ready, but It is still marked as WIP, since sklearn/scipy install failed and I did not have the time to investigate this further so far...

**Comment by Borda** [comment]
> @Borda this should be ready, but It is still marked as WIP, since sklearn/scipy install failed and I did not have the time to investigate this further so far...

ok, lest make is ready and I ll drop wip to be clear that with passing test we can merge it :rabbit:

**Comment by Borda** [comment]
@justusschock do we need `imgmath_latex`? can you point out where it is used?

**Comment by awaelchli** [comment]
@Borda I think I added it when I was requested for review on metrics because there was a formula and it did not look right. PyTorch also uses imgmath latex. Is it creating problems?

File: pytorch_lightning/metrics/sklearn.py [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+                Defaults to sum.
+            **kwargs: additonal keyword arguments (will be forwarded to metric call)
+        """
+        super().__init__(name=metric_name, reduce_group=reduce_group,
+                         reduce_op=reduce_op)
+
+        self.metric_kwargs = kwargs
+
+        lightning_logger.debug(
+            'Every metric call will cause a GPU synchronization, which may slow down your code')
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/sklearn.py:53]: Should this be maybe be changed to `lightning_logger.warning`? 
Also I would suggest that we made the text metric specific:
```suggestion
            f'Metric {self.__class__.__name__} is using Sklearn as backend, meaning that'
            ' every metric call will cause a GPU synchronization, which may slow down your code'
            
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -81,7 +84,7 @@ jobs:
 
   Build-Docs:
     docker:
-      - image: circleci/python:3.7
```
Comment (awaelchli) [inline .circleci/config.yml:84]: not sure, maybe this is causing the docs build error. 
btw why this change?

Comment (Borda) [reply .circleci/config.yml:84]: because `circleci/python:3.7` does not have `latex` and latex is needed for imgmath extension

Comment (Borda) [reply .circleci/config.yml:84]: so the problem is in python version, py3.6 does not support the inheritance... the min version is 3.7

**Comment by Borda** [comment]
> @Borda I think I added it when I was requested for review on metrics because there was a formula in the docs and it did not look right, so I fixed it. PyTorch also uses imgmath latex. Is it creating problems?

kind of, because of `imgmath` extension, I had to pull an image with latex instead of the general one

File: pytorch_lightning/metrics/sklearn.py [hunk 6]
```diff
@@ -0,0 +1,716 @@
...
+        Args:
+            y_pred: the array containing the predictions (already in categorical form)
+            y_true: the array containing the targets (in categorical form)
+            sample_weight:  Sample weights.
+
+        Return:
+            Accuracy Score
+
+        """
+        return super().forward(y_pred=y_pred, y_true=y_true, sample_weight=sample_weight)
```
Comment (tullie) [inline pytorch_lightning/metrics/sklearn.py:114]: Do you need any of these forward functions in this file? It seems like you're implementing what inheritance will already do, i.e. call the superfunction if it's not overriden.

File: pytorch_lightning/metrics/utils.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    Args:
+        data: the data to convert to tensor
+
+    Returns:
+        the converted data
+
+    """
+    if isinstance(data, numbers.Number):
+        return torch.tensor([data])
+
```
Comment (tullie) [inline pytorch_lightning/metrics/utils.py:47]: nit: remove this whiteline

File: pytorch_lightning/metrics/sklearn.py (hunk 6 above)
Comment (Borda) [reply pytorch_lightning/metrics/sklearn.py:114]: @justusschock ^^

Comment (justusschock) [reply pytorch_lightning/metrics/sklearn.py:114]: @tullie you're right. But the superfunction takes args and kwargs as arguments. I have explicitly implemented these for documentation and type hinting purposes. But we can also remove them.
